pkg/apis/apps/validation/validation.go

```diff
@@ -75,6 +75,48 @@ func ValidateStatefulSetSpec(spec *apps.StatefulSetSpec, fldPath *field.Path) fi
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("podManagementPolicy"), spec.PodManagementPolicy, fmt.Sprintf("must be '%s' or '%s'", apps.OrderedReadyPodManagement, apps.ParallelPodManagement)))
 	}
 
+	switch spec.UpdateStrategy.Type {
+	case "":
+		allErrs = append(allErrs, field.Required(fldPath.Child("updateStrategy"), ""))
+	case apps.OnDeleteStatefulSetStrategyType, apps.RollingUpdateStatefulSetStrategyType:
+		if spec.UpdateStrategy.Partition != nil {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath.Child("updateStrategy").Child("partition"),
+					spec.UpdateStrategy.Partition.Ordinal,
+					fmt.Sprintf("only allowed for updateStrategy '%s'", apps.PartitionStatefulSetStrategyType)))
+		}
+	case apps.PartitionStatefulSetStrategyType:
+		if spec.UpdateStrategy.Partition == nil {
+			allErrs = append(
+				allErrs,
+				field.Required(
+					fldPath.Child("updateStrategy").Child("partition"),
+					fmt.Sprintf("required for updateStrategy '%s'", apps.PartitionStatefulSetStrategyType)))
+			break
+		}
+
+		allErrs = append(allErrs,
+			apivalidation.ValidateNonnegativeField(
+				int64(spec.UpdateStrategy.Partition.Ordinal),
+				fldPath.Child("updateStrategy").Child("partition").Child("ordinal"))...)
+		if spec.UpdateStrategy.Partition.Ordinal > spec.Replicas {
+			allErrs = append(
+				allErrs, field.Invalid(
+					fldPath.Child("updateStrategy").Child("partition").Child("ordinal"),
+					spec.UpdateStrategy.Partition.Ordinal,
+					"must be less than 'replicas'"))
```

Comment:
the check above suggests the partition can be equal to replicas but the error says otherwise. We need to fix the check above